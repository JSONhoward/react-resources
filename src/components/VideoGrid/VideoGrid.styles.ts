import styled from "styled-components";

export const StyledVideoGrid = styled('div')`
width: 100%;
padding: 1rem;
`

export const Videos = styled('div')`
display: grid;
grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
justify-items: center;
gap: 2rem;
`

export const SingleVideo = styled('div')`
position: relative;
display: flex;
align-items: center;
justify-content: center;
height: fit-content;
overflow: hidden;
box-shadow: 1px 1px 5px rgba(0,0,0,.5);

img {
    height: auto;
    width: 100%;
    object-fit: cover;
}
`

export const VideoImagePlay = styled('div')`
position: absolute;
height: 100%;
width: 100%;
backdrop-filter: blur(2px);
overflow: hidden;
opacity: 0;
transition: all .5s ease-in-out;
cursor: pointer;

&:hover {
    opacity: 1;
}

svg {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}

span {
    position: absolute;
    top: .5rem;
    left: 50%;
    transform: translateX(-50%);
    color: white;
    text-shadow: 1px 1px 5px black;
}
`