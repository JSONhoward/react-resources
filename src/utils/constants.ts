import cleverP from '../images/clever-programmer-icon.jpg'
import devEd from '../images/dev-ed-icon.jpg'
import fireship from '../images/fireship-icon.jpg'
import freeCodeCamp from '../images/free-code-camp-icon.jpg'
import netNinja from '../images/net-ninja-icon.jpg'
import webDevSimplified from '../images/web-dev-simplified-icon.jpg'
import rwieruch from '../images/rwieruch-icon.svg'
import sitePoint from '../images/sitepoint.svg'
import smashingMagazine from '../images/smashing-magazine.svg'
import codeCademy from '../images/codecademy.svg'
import udemy from '../images/udemy-icon.svg'
import wesBos from '../images/wes-bos-icon.png'
import medium from '../images/medium-icon.png'
import devTo from '../images/dev-to.svg'

export const webDevChannels: { name: string, id: string, icon: string }[] = [
    {
        id: 'UCFbNIlppjAuEX4znoulh0Cw',
        name: 'Web Dev Simplified',
        icon: webDevSimplified
    },
    {
        id: 'UClb90NQQcskPUGDIXsQEz5Q',
        name: 'Dev Ed',
        icon: devEd
    },
    {
        id: 'UCW5YeuERMmlnqo4oq8vwUpg',
        name: 'Net Ninja',
        icon: netNinja
    },
    {
        id: 'UCsBjURrPoezykLs9EqgamOA',
        name: 'Fireship',
        icon: fireship
    },
    {
        id: 'UC8butISFwT-Wl7EV0hUK0BQ',
        name: 'Free Code Camp',
        icon: freeCodeCamp
    },
    {
        id: 'UCqrILQNl5Ed9Dz6CGMyvMTQ',
        name: 'Clever Programmer',
        icon: cleverP
    }
]

export const mockArray: {id: string, url: string, title: string, description: string, thumb: string}[] = [
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    },
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }, {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }, {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    },
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }
]

export const onlineCourses: {platform: string, name: string, url: string, icon: string}[] = [
    {
        platform: 'FreeCodeCamp',
        name: 'Learn React in 1 Hour by Building a Movie Search App',
        url: 'https://www.freecodecamp.org/news/learn-react-in-1-hour-by-building-a-movie-search-app/',
        icon: freeCodeCamp
    },
    {
        platform: 'Udemy',
        name: 'React - The Complete Guide (incl Hooks, React Router, Redux)',
        url: 'https://www.udemy.com/course/react-the-complete-guide-incl-redux/',
        icon: udemy
    },
    {
        platform: 'CodeCademy',
        name: 'Learn React.js: Part I/Part II',
        url: 'https://www.codecademy.com/learn/react-101',
        icon: codeCademy
    },
    {
        platform: 'none',
        name: 'Fullstack Advanced React & GraphQL',
        url: 'https://advancedreact.com/',
        icon: wesBos
    }
]

export const onlineBlogs: {name: string, url: string, icon: string}[] = [
    {
        name: 'Medium',
        url: 'https://medium.com/',
        icon: medium
    },
    {
        name: 'Dev.to',
        url: 'https://dev.to/',
        icon: devTo
    },
    {
        name: 'Sitepoint',
        url: 'https://www.sitepoint.com/',
        icon: sitePoint
    },
    {
        name: 'Smashing Magazine',
        url: 'https://www.smashingmagazine.com/',
        icon: smashingMagazine
    },
    {
        name: 'RobinWieruch',
        url: 'https://www.robinwieruch.de/blog',
        icon: rwieruch
    }
]