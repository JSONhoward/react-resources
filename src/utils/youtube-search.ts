const YOUTUBE_SEARCH_URL = (channel: string, limit: string): string => `https://www.googleapis.com/youtube/v3/search?part=snippet&channelId=${channel}${limit}&q=react&type=video&key=${process.env.REACT_APP_YOUTUBE_API}`

export const getVideoList = async (channel: string, limit = 50): Promise<Video[]> => {
    const response = await fetch(YOUTUBE_SEARCH_URL(channel, `&maxResults=${limit.toString()}`))
    const data = await response.json()

    const videos: Video[] = await data['items'].map((result: { id: { videoId: any }; snippet: { title: any; description: any; thumbnails: { high: { url: any } } } }) => {
        return {
            id: result.id.videoId,
            url: `http://www.youtube.com/watch?v=${result.id.videoId}`,
            title: result.snippet.title,
            description: result.snippet.description,
            thumb: result.snippet.thumbnails.high.url
        }
    })

    return videos
}