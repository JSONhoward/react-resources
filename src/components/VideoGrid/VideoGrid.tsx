import React from 'react'
import { FaPlay } from 'react-icons/fa';
import { mockArray } from '../../utils/constants';
// import { getVideoList } from '../../utils/youtube-search';

import { StyledVideoGrid, SingleVideo, Videos, VideoImagePlay } from './VideoGrid.styles';

type Props = {
    openPortal: (id: string) => void,
    channelId: string
}

const VideoGrid: React.FC<Props> = ({ channelId, openPortal }) => {
    const [videoImages, setVideoImages] = React.useState<Video[] | undefined>(undefined)

    // React.useEffect(() => {
    //     getVideoList(channelId, 10)
    //     .then(videos => setVideoImages(videos))
    //     .catch(error => console.error('Error: ', error))
    // }, [])

    React.useEffect(() => {
        setVideoImages(mockArray)
    },[])

    const videoList = videoImages?.map((el, i) => {
        return (
            <SingleVideo onClick={() => openPortal(el.url)} key={i}>
                <VideoImagePlay>
                    <span>{el.title}</span>
                    <FaPlay size={'3rem'} color={'white'} />
                </VideoImagePlay>
                <img data-testid={`singleVideoImage${i}`} src={el.thumb} alt={el.title} />
            </SingleVideo>
        )
    })

    return (
        <StyledVideoGrid>
            <Videos>
                {videoList}
            </Videos>
        </StyledVideoGrid>
    )
}

export default VideoGrid
