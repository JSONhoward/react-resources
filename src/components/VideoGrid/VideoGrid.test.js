import React from 'react'
import {create} from 'react-test-renderer'
import {screen, render} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import VideoGrid from './'
import { mockArray } from '../../utils/constants'

describe('VideoGrid Tests', () => {
    test('Snapshot Test', () => {
        const tree = create(<VideoGrid />).toJSON()

        expect(tree).toMatchSnapshot()
    })

    // test('Renders Props', () => {
    //     const testChannelId = 'test id',
    //         testOpenPortal = jest.fn(() => console.log('open portal'))

    //     render(<VideoGrid channelId={testChannelId} openPortal={testOpenPortal} />)
    // })

    test('Renders list of videos', () => {
        render(<VideoGrid />)

        mockArray.forEach((el, i) => {
            expect(screen.getByTestId(`singleVideoImage${i}`)).toHaveAttribute('src', el.thumb)
            expect(screen.getByTestId(`singleVideoImage${i}`)).toHaveAttribute('alt', el.title)
            expect(screen.getByText(el.title)).toBeInTheDocument()
        })
    })
})