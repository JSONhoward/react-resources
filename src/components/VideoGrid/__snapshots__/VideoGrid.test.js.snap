// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoGrid Tests Snapshot Test 1`] = `
<div
  className="sc-bdnylx CtqKO"
>
  <div
    className="sc-gtssRu hyjofv"
  />
</div>
`;
